### PR TITLE
[Notifier] Rework the Smsbox bridge

### PR DESCRIPTION
**Pull request by alexandre-daubois**
| Q             | A
| ------------- | ---
| Branch?       | 7.1
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Issues        | -
| License       | MIT

Rework the whole component to match the rest of the codebase.

### CODE REVIEW

File: src/Symfony/Component/Notifier/Bridge/Smsbox/SmsboxOptions.php [hunk 1]
```diff
@@ -250,7 +250,6 @@ public static function validateDate(string $date): string
 
     public static function validateDateTime(\DateTime $dateTime)
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Notifier/Bridge/Smsbox/SmsboxOptions.php]: we should never use DateTime anymore so we should change all occurences to DateTimeImmutable

File: src/Symfony/Component/Notifier/Bridge/Smsbox/SmsboxOptions.php [hunk 2]
```diff
@@ -250,7 +250,6 @@ public static function validateDate(string $date): string
...
     {
-        \Locale::setDefault('fr');
         $now = new \DateTime();
         $tz = new \DateTimeZone('Europe/Paris');
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Notifier/Bridge/Smsbox/SmsboxOptions.php]: this is super strange validation logic

File: src/Symfony/Component/Notifier/Bridge/Smsbox/SmsboxOptions.php (hunk 1 above)
Comment (nicolas-grekas) [reply src/Symfony/Component/Notifier/Bridge/Smsbox/SmsboxOptions.php]: also, all public static method should be private to me

**Review by nicolas-grekas — COMMENTED** [review]
looks like there are non-strict comparisons in the bridge
overall, it looks like the component needs a review to make it up to par with the rest of the codebase, could you have a look?

File: src/Symfony/Component/Notifier/Bridge/Smsbox/SmsboxOptions.php [hunk 3]
```diff
@@ -11,375 +11,203 @@
...
-        $this->options = [];
     }
 
     public function getRecipientId(): null
     {
         return null;
     }
 
-    public function mode(string $mode)
+    public function mode(SmsboxMessageMode $mode): self
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Notifier/Bridge/Smsbox/SmsboxOptions.php]: `@return $this` also where appropriate

File: src/Symfony/Component/Notifier/Bridge/Smsbox/SmsboxOptions.php [hunk 4]
```diff
@@ -11,375 +11,203 @@
...
+    public function dateTime(\DateTimeImmutable $dateTime): self
     {
-        $this->options['charset'] = self::validateCharset($charset);
+        if (isset($this->options['date']) || isset($this->options['heure'])) {
+            throw new InvalidArgumentException(sprintf('Either %1$s::dateTime() or %1$s::date() and %1$s::hour() must be called, but not both.', self::class));
+        }
 
-        return $this;
-    }
+        if ($dateTime < now()) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Notifier/Bridge/Smsbox/SmsboxOptions.php]: let's not add a dep on Clock just for validation

File: src/Symfony/Component/Notifier/Bridge/Smsbox/composer.json [hunk 5]
```diff
@@ -25,7 +25,9 @@
     ],
     "require": {
         "php": ">=8.2",
+        "symfony/clock": "^6.4|^7.0",
         "symfony/http-client": "^6.4|^7.0",
+        "symfony/intl": "^6.4|^7.0",
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Notifier/Bridge/Smsbox/composer.json]: why do we need this one?

Comment (alexandre-daubois) [reply src/Symfony/Component/Notifier/Bridge/Smsbox/composer.json]: To use the `Countries` class in SmsboxOptions, this ease the check of the country code validity

File: src/Symfony/Component/Notifier/Bridge/Smsbox/SmsboxOptions.php [hunk 6]
```diff
@@ -11,375 +11,203 @@
...
+    public function hour(string $hour): self
     {
-        $this->options['variable'] = $variable;
+        if (isset($this->options['dateTime'])) {
+            throw new InvalidArgumentException(sprintf('Either %1$s::dateTime() or %1$s::date() and %1$s::hour() must be called, but not both.', self::class));
+        }
 
-        return $this;
-    }
+        if (false === \DateTimeImmutable::createFromFormat('H:i', $hour)) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Notifier/Bridge/Smsbox/SmsboxOptions.php]: same everywhere similar:
```suggestion
        if (!\DateTimeImmutable::createFromFormat('H:i', $hour)) {
```

File: src/Symfony/Component/Notifier/Bridge/Smsbox/SmsboxOptions.php [hunk 7]
```diff
@@ -11,375 +11,203 @@
...
+            throw new InvalidArgumentException('Hour must be in HH:MM format.');
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Notifier/Bridge/Smsbox/SmsboxOptions.php]: `\DateMalformedStringException` with the 8.3 polyfill?

File: src/Symfony/Component/Notifier/Bridge/Smsbox/composer.json (hunk 5 above)
Comment (nicolas-grekas) [reply src/Symfony/Component/Notifier/Bridge/Smsbox/composer.json]: let's make it optional and skip the validation is the class isn't found?

**Comment by alexandre-daubois** [comment]
Comments are addressed. I'll regenerate the expected return types file in a moment 🙂

**Review by nicolas-grekas — APPROVED** [review]

**Comment by alexandre-daubois** [comment]
Changed `self` return types to `static` in `SmsboxOptions` which resolved the return type issues. This should be ready now

**Comment by OskarStark** [comment]
Great work, I would prefer to remove the SmsboxMessage prefix from the enums, we have the namespace.

Agree @nicolas-grekas ?

**Comment by nicolas-grekas** [comment]
`Enum` subnamespace?

**Comment by alexandre-daubois** [comment]
I was skeptical to just have a `Day` or `Charset` enum for instance at the root of the component, but with an Enum subnamespace, I agree with the idea. I'm doing the change

**Comment by alexandre-daubois** [comment]
Here you go, enums have been moved to their namespace

**Comment by fabpot** [comment]
Thank you @alexandre-daubois.

**Pull request merged**